### PR TITLE
Run npm install as part of the install (fixes #12519)

### LINKED ISSUE #12519: run npm install in make install to pull in html-minifier-terser
I am just trying out emscripten and when I use emcc -O2 -o a.html (as in [this example](https://emscripten.org/docs/porting/asyncify.html#returning-values)) I get the following error:

`emcc: error: html-minifier-terser was not found! Please run "npm install" in Emscripten root directory to set up npm dependencies`.

I have installed emscripten using the Arch Linux [package](https://github.com/archlinux/svntogit-community/blob/packages/emscripten/trunk/PKGBUILD#L97), and experience suggests that it would be unwise to run npm install there. I contacted the package maintainer and was recommended to report this issue upstream instead, as make install should definitely also take care of that step. So, could the necessary `npm install` be added to `make install`?

### PR DESCRIPTION
**Pull request by svenstaro**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -7,6 +7,7 @@ dist: $(DISTFILE)
 install:
 	@rm -rf $(DESTDIR)
 	./tools/install.py $(DESTDIR)
+	npm install --prefix $(DESTDIR)
```
Comment (sbc100) [inline Makefile:10]: Is that the same as doing `cd $(DESTDIR) && npm install`?

**Comment by sbc100** [comment]
Should we also add `EXCLUDE_PATTERN=--exclude=node_modules` in order to avoid shipping node_modules in that tar archive?

**Review by sbc100 — COMMENTED** [review]
Looking at `./tools/install.py` it looks like it deliberatly excludes node_modules.

I wonder if we should have `tools/install.py` run `npm install` first and not exclude node_modules?

When we could re-enable `EXCLUDE_PATTERN=--exclude=node_modules` in the Makefile which should mean node_modules appears in the installation directory, but not in the tar ball.

File: Makefile (hunk 1 above)
Comment (svenstaro) [reply Makefile:10]: Yes.

**Comment by svenstaro** [comment]
I think it makes sense to ship node_modules, by the way. Then users don't need to have `npm` installed. Since you're shipping the `package-lock.json` file they should end up with the exact versions from the lockfile anyway.

**Comment by sbc100** [comment]
> I think it makes sense to ship node_modules, by the way. Then users don't need to have `npm` installed. Since you're shipping the `package-lock.json` file they should end up with the exact versions from the lockfile anyway.

I don't think node_modules is cross-platform though.   `npm install` can perform complication using using local compiler (gcc or msvc or whatever) which means the resulting tree can contain not just OS-specific but distro and release-specific binaries.   If it was guaranteed to be pure JS I would tend to agree with you.  The release tarball is supposed to be cross-platform source distribution.

**Comment by svenstaro** [comment]
I see. Yeah, that makes sense. I'm not really sure what makes the most sense for you here. I only know that I need node_modules to exist for my packages and that it probably makes sense for a regular user typing `make install` as well.

**Comment by sbc100** [comment]
> I see. Yeah, that makes sense. I'm not really sure what makes the most sense for you here. I only know that I need node_modules to exist for my packages and that it probably makes sense for a regular user typing `make install` as well.

What about having `make install` run `npm install` but `make dist` exclude that directory?

**Comment by svenstaro** [comment]
Yeah, that's a fine suggestion that makes sense to me.

File: Makefile [hunk 2]
```diff
@@ -7,11 +7,12 @@ dist: $(DISTFILE)
...
 
 # Create an distributable archive of emscripten suitable for use
 # by end users.  This archive excludes parts of the codebase that
 # are you only used by emscripten developers.
 $(DISTFILE): install
-	tar cf $@ $(EXCLUDE_PATTERN) -C `dirname $(DESTDIR)` `basename $(DESTDIR)`
+	tar cf $@ $(EXCLUDE_PATTERN) --exclude=node_modules -C `dirname $(DESTDIR)` `basename $(DESTDIR)`
```
Comment (sbc100) [inline Makefile:16]: I think this is what `EXCLUDE_PATTERN` used to be for.  Can you use (or just remove it?)?

Comment (svenstaro) [reply Makefile:16]: I didn't find a way for EXCLUDE_PATTERN to be used actually and assumed it was some magic I didn't understand. I'll be happy to take it out with your approval, though.

Comment (sbc100) [reply Makefile:16]: This file used to to have `EXCLUDE_PATTERN=--exclude='*.pyc' --exclude='*/__pycache__'` up top before #10837.

Comment (svenstaro) [reply Makefile]: In that case, it should be fine now. :)

**Review by sbc100 — APPROVED** [review]

**Comment by svenstaro** [comment]
CI appears to be broken here as well. What am I doing wrong?

**Comment by sbc100** [comment]
Oh odd... normally circleci runs just run with other contribute PRs from forked repos.

Can you click in the circleci link from one of the other builds?  I would be surprised if you need to log into circleci because the CI should be running under our account...

**Comment by svenstaro** [comment]
I now granted Circle CI explicit permissions. Let's see...

**Comment by svenstaro** [comment]
@sbc100 Finally I got this to run on CI!

**Review by sbc100 — APPROVED** [review]

**Pull request merged**